### ISSUE TITLE
RFC Declare @throws in phpdoc

### ISSUE THREAD

**Issue by tractorcow** [silverstripe/silverstripe-framework#8151]
This suggestion discusses the adding of standard `@throws` phpdoc to any method that is known to throw exceptions (or calls methods that throw these exceptions).

Pros:
 - better visibility to users of what exceptions code can throw.
 - Ability to track semver violations where exception types change

Cons:
 - If we adopt this, we'll need to essentially add these everywhere we throw exceptions in a big PR.
 - We'll need to maintain this going forward also

I suggest that we vote to either go all in on this, or agree to avoid adding this to core.

cc @silverstripe/core-team for discussion.

**Comment by kinglozzer** [comment]
Strong +1 for adding these

**Comment by flamerohr** [comment]
> or calls methods that throw these exceptions
I'm hoping it's a "common sense" type thing here, otherwise this could turn into a maintenance nightmare

**Comment by robbieaverill** [comment]
+1 for documenting exceptions throw directly by methods. We can't document exceptions thrown by dependant methods really, otherwise everything would probably have an annotation

**Comment by maxime-rainville** [comment]
I'm in favor, but how does this impacts our commitment to `semver`. Do we consider this part of our official API?

If it is, then you better think things through before saying "this method throws `Exception X`",  because you've got to wait until the next major release to change it if you change your mind.

**Comment by unclecheese** [comment]
The contract we have with semver, as I understand it, is to not break anyone's code. This change might affect developer ergonomics, but I don't see how it could break functional code.

**Comment by robbieaverill** [comment]
Yeah. Documenting existing exceptions doesn't affect semver. Adding or removing actual exceptions would though.

**Comment by tractorcow** [comment]
ok, just direct methods then. ;)

**Comment by sminnee** [comment]
Is there a linter / static analyser that can help us with this?

**Comment by maxime-rainville** [comment]
```xml
<?xml version="1.0" encoding="UTF-8"?>
<ruleset name="SilverStripe">
    <rule ref="Squiz.Commenting.FunctionCommentThrowTag">
    ...
    </rule>
</ruleset>
```

**Comment by dhensby** [comment]
> Is there a linter / static analyser that can help us with this?

PHPStorm does do this out of the box, but I'm not sure what linters replicate that

**Comment by wernerkrauss** [comment]
> Is there a linter / static analyser that can help us with this?

https://plugins.jetbrains.com/plugin/7622-php-inspections-ea-extended- is a very good and popular plugin that helps me finding this kind of documentation error (and other stuff) in PHPStorm.

It can afaik also fix the missing `@throws` automatically.

**Comment by robbieaverill** [comment]
`rfc/accepted` for ensuring that all methods that _directly_ throw exceptions (rather than methods they call e.g. `parent::foo()` or `$this->bar()` throwing them)?

**Comment by robbieaverill** [comment]
No feedback in 6 weeks, removing feedback required label

**Comment by sminnee** [comment]
> Is there a way to enable it without forcing us to immediately fix them every where?

How much time would it take to add it everywhere in a single package?

**Comment by maxime-rainville** [comment]
Not sure if this is the best way of doing this, but travis has a TRAVIS_COMMIT_RANGE variable. We could use that to [pick out only the files that have changed on the current PR build](https://stackoverflow.com/a/38550849/1427439) and run some stricter tests on those files only.
